src/core/expr/fexpr_fillna.cc

```diff
@@ -58,56 +60,17 @@ namespace dt {
 
         Workframe evaluate_n(EvalContext &ctx) const override {
           Workframe wf = arg_->evaluate_n(ctx);
-          // Groupby gby = Groupby::single_group(wf.nrows());
-
-          // if (ctx.has_groupby()) {
-          //   wf.increase_grouping_mode(Grouping::GtoALL);
-          //   gby = ctx.get_groupby();
-          // }
-
-          // bool bool_direction = direction_=="down";
-          // bool is_valid;
-          // int64_t res;
-          // for (size_t i = 0; i < wf.ncols(); ++i) {
-          //   Column coli = wf.retrieve_column(i);
-          //   is_valid = coli.get_element(4, &res);
-          //   std::cout<<is_valid<<std::endl;
-          //   std::cout<<res<<std::endl;
-          //   //Column coli = evaluate1(wf.retrieve_column(i), bool_direction, gby);
-          //   //wf.replace_column(i, std::move(coli));
-          // }
-          return wf;
+          Workframe out(ctx);
+          for (size_t i = 0; i < wf.ncols(); ++i) {
+            Column coli = wf.retrieve_column(i);
+            coli = Column(new Isna_ColumnImpl<int8_t>(std::move(coli)));
+            RowIndex ri = RowIndex(std::move(coli));
+            coli.apply_rowindex(ri.negate(coli.nrows())); // i get bools here
+            out.add_column(std::move(coli), std::string(), out.get_grouping_mode());
```

Comment:
What you need to do here is to go through all the elements in the column `coli` and depending on their validity fill the RowIndex `ri`:
- if `i`-th element is valid, then `ri[i]` should be set to `i`;
- if `i`-th element is not valid, then `ri[i]` should be set to the index of the last valid element.

At the end you need to apply `ri` to the original column `coli`.

Reply:
@oleksiyskononenko I am getting this error message : 
```
src/core/expr/fexpr_fillna.cc:73:17: error: type 'RowIndex' does not provide a subscript operator
              ri[j] = prev;
```

Reply:
Yeah, the subscript operator is defined for the underlying `Buffer` only. See, for example, how it is done for the `shift()` function: https://github.com/h2oai/datatable/blob/main/src/core/expr/head_func_shift.cc#L43-L44

Once the `Buffer` is filled out, you simply create a `RowIndex` as https://github.com/h2oai/datatable/blob/main/src/core/expr/head_func_shift.cc#L64

Reply:
@oleksiyskononenko updated. is this what you have in mind?

Reply:
Good job, @samukweku. Yes, we're pretty close now. Let's just fix the tests, because they seems to use the old `direction` variable instead of the `reverse` one. Once the tests are fixed, we will see the real issues if any.

Reply:
all done. I need your expertise on this @oleksiyskononenko ... I am getting segmentation fault - it has to do with memory access but I can't figure it out, specifically lines 117 - 128. I am trying to implement fillna if a value is supplied; I am using the if else implementation for this.

Reply:
I suggest we start with `reverse` parameter only and once this PR is green, we merge it. Then, we gradually add other options to `fillna()`. For the moment, it seems that after you fixed the tests the PR is still red, meaning that we have some issues with even the initial implementation.

Reply:
@oleksiyskononenko done

Reply:
Thanks, one test is still failing though

```
FAILED tests/test-docs.py::test_documentation_h - AssertionError: Variable doc_FExpr_fillna declared in documentation.h not found as a :cvar: among *.rst files
```

Reply:
All done; speed looks ok; however, there are some scenarios where `by` is quite slow compared to pandas and affects the outcome - in fact based on a naive observation, most of the time is spent in the grouping. I'm hoping we can have a look at the by internals sometime in the future and see if there are any low hanging fruits than can help improve the speed. #2551.       #2541

Reply:
Thanks, let me review the PR then.